src/chainstate/stacks/address.rs

```diff
@@ -21,24 +21,238 @@ use net::StacksMessageCodec;
 use net::Error as net_error;
 use net::codec::{read_next, write_next};
 
+use address::AddressHashMode;
+use address::c32::c32_address;
+use address::public_keys_to_address_hash;
+
+use chainstate::stacks::StacksPublicKey;
 use chainstate::stacks::StacksAddress;
+use chainstate::stacks::STACKS_ADDRESS_ENCODED_SIZE;
 
 use util::hash::Hash160;
+use util::hash::HASH160_ENCODED_SIZE;
+
+use burnchains::PublicKey;
+use burnchains::Address;
+
+use address::c32::c32_address_decode;
+
+use burnchains::bitcoin::address::BitcoinAddress;
+use burnchains::bitcoin::address::{
+    address_type_to_version_byte,
+    to_c32_version_byte
+};
+
+use vm::types::{
+    PrincipalData,
+    StandardPrincipalData,
+};
+
 
 impl StacksMessageCodec for StacksAddress {
     fn serialize(&self) -> Vec<u8> {
         let mut res = vec![];
         write_next(&mut res, &self.version);
-        write_next(&mut res, &self.bytes.as_bytes().to_vec());
+        res.extend_from_slice(self.bytes.as_bytes());
         res
     }
 
-    fn deserialize(buf: &Vec<u8>, index: &mut u32, max_size: u32) -> Result<StacksAddress, net_error> {
-        let version : u8    = read_next(buf, index, max_size)?;
-        let bytes : Hash160 = read_next(buf, index, max_size)?;
+    fn deserialize(buf: &Vec<u8>, index_ptr: &mut u32, max_size: u32) -> Result<StacksAddress, net_error> {
+        let mut index = *index_ptr;
+
+        let new_index = index.checked_add(STACKS_ADDRESS_ENCODED_SIZE).ok_or(net_error::DeserializeError)?;
```

Comment:
What is `new_index` used for? Is it only to bounds check the additions further down? If so, it'd be better to do the bounds check on the final indexes directly. I think there's an off-by-one error as the checks are currently written -- `new_index` is checked for > buf.len, but it should be an overflow if its >= buf.len.

Reply:
Good catch -- that should be >=.  `new_index` is just used for the bounds-check.  However, `index` will be (implicitly) advanced to the value of `new_index` here because we use it to read a `u8` and a `Hash160`, or 21 bytes (which is the same value as `STACKS_ADDRESS_ENCODED_SIZE`.  I can make this clearer.

Reply:
Fixed in 3cbc89e1